### PR TITLE
feat(machinery): initial work for machinery

### PR DESCRIPTION
**Pull request by thesayyn**

### CODE REVIEW

File: stacks/api/bucket/src/bucket.module.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
-import {DynamicModule, Module, Type, Global} from "@nestjs/common";
+import {DynamicModule, Global, Module, Type} from "@nestjs/common";
```
Comment (htuna07) [inline stacks/api/bucket/src/bucket.module.ts:1]: Can we move these logics to a more specific file for each module?

File: stacks/api/machinery/src/store.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  async set(name: string, value: Resource<SpecType, StatusType>): Promise<void> {
+    value["_id"] = `${this.groupKey}ɵ${name}`;
+    await this.store.replaceOne({_id: value["_id"]}, value, {
+      upsert: true
+    });
+  }
+
+  async patch(name: string, patch: DeepPartial<Resource<SpecType, StatusType>>) {
```
Comment (htuna07) [inline stacks/api/machinery/src/store.ts:50]: We have the same method on bucket utilities. Can we move that `getUpdateQueryForPatch` method to the common package and use it?

File: stacks/api/machinery/src/validation.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+//             oneOf.message = `either ${oneOfRelatedErrors.map(error => error.message).join(" or ")}`;
+//             errors = errors.filter((error, index) => error.dataPath != oneOf.dataPath || index == oneOfIndex);
+//         } else {
+//             console.log(oneOf);
+//         }
+
+//     }
+
+//     return errors;
+// }
```
Comment (htuna07) [inline stacks/api/machinery/src/validation.ts:64]: Can we add one more comment line that explains why we disabled these lines or uncomment all of them?

File: stacks/cli/src/commands/delete.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import {ActionParameters, CaporalValidator, Command, CreateCommandParameters} from "@caporal/core";
+import * as fs from "fs";
+import * as path from "path";
+import * as YAML from "yaml";
+import {machinery} from "../machinery";
+import {formatFailureStatus, isFailureStatus} from "../status";
+
+async function del({options}: ActionParameters) {
```
Comment (htuna07) [inline stacks/cli/src/commands/delete.ts:8]: `remove` is more understandable than `del` I think.

File: stacks/spica/src/dashboard/pages/tutorial/tutorial.component.html [hunk 5]
```diff
@@ -1,6 +1,88 @@
 <div class="content-container">
   <div class="logo">
-    <img src="assets/logo-1.png" />
+    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 111.1 111.1">
```
Comment (htuna07) [inline stacks/spica/src/dashboard/pages/tutorial/tutorial.component.html:3]: We have a pr that fixes this bug. Can we revert this change?

File: stacks/api/activity/services/src/activity.interceptor.ts [hunk 6]
```diff
@@ -43,7 +43,7 @@ export function getUser(user: any): string {
   return user ? user._id : undefined;
 }
 
-export function createActivity(
+export async function createActivity(
```
Comment (htuna07) [inline stacks/api/activity/services/src/activity.interceptor.ts:46]: Returning a promise is better than make this function void I think. Why did we need to make this change?

File: stacks/api/activity/services/src/activity.service.ts [hunk 7]
```diff
@@ -8,8 +8,8 @@ const COLLECTION_NAME = "activity";
 export class ActivityService extends BaseCollection<Activity>(COLLECTION_NAME) {
   constructor(db: DatabaseService, @Inject(ACTIVITY_OPTIONS) options: ActivityOptions) {
     super(db);
-    this.createCollection(COLLECTION_NAME).then(() =>
-      this.upsertTTLIndex(options.expireAfterSeconds)
-    );
+    this.createCollection(COLLECTION_NAME)
+      .then(() => this.upsertTTLIndex(options.expireAfterSeconds))
+      .catch(() => {});
```
Comment (htuna07) [inline stacks/api/activity/services/src/activity.service.ts:13]: Can we use `console.log` for this error instead of ignoring them?

File: stacks/api/activity/services/src/activity.interceptor.ts (hunk 6 above)
Comment (thesayyn) [reply stacks/api/activity/services/src/activity.interceptor.ts:46]: I have made it this way because I don't want this function to return the result of the insertMany.

File: stacks/spica/src/dashboard/pages/tutorial/tutorial.component.html (hunk 5 above)
Comment (thesayyn) [reply stacks/spica/src/dashboard/pages/tutorial/tutorial.component.html:3]: done

File: stacks/cli/src/commands/delete.ts (hunk 4 above)
Comment (thesayyn) [reply stacks/cli/src/commands/delete.ts:8]: replaced with `_delete`. I could not use delete cause it is reserved.

File: stacks/api/machinery/src/validation.ts (hunk 3 above)
Comment (thesayyn) [reply stacks/api/machinery/src/validation.ts:64]: This was a part of an experimental feature that I have failed to implement. Better remove it

File: stacks/api/machinery/src/store.ts (hunk 2 above)
Comment (thesayyn) [reply stacks/api/machinery/src/store.ts:50]: This an implementation detail of the store functionality of machinery. TBH I don't wanna tie it to a common function since this might change in the future. I think for now we better not couple it with any other patching mechanism in the codebase.

Comment (thesayyn) [reply stacks/api/machinery/src/store.ts:50]: This might change in the future with an extended version of merge+patch that we are going to implement.

File: stacks/api/machinery/src/validation.ts (hunk 3 above)
Comment (thesayyn) [reply stacks/api/machinery/src/validation.ts:64]: Done

File: stacks/api/activity/services/src/activity.interceptor.ts (hunk 6 above)
Comment (thesayyn) [reply stacks/api/activity/services/src/activity.interceptor.ts:46]: since it is an async function it returns `Promise<void>` since I didn't return anything.

File: stacks/api/machinery/src/store.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        const key = base ? `${base}.${name}` : name;
+        const value = partialPatch[name];
+        const type = typeof value;
+
+        if (value == null) {
+          unset[key] = "";
+        } else if (type == "boolean" || type == "string" || type == "number" || type == "bigint") {
+          set[key] = value;
+        } else if (Array.isArray(value)) {
+          throw new Error("TODO: handle arrays");
```
Comment (htuna07) [inline stacks/api/machinery/src/store.ts:65]: We can use `set[key] = value` instead of throwing an error by the way.

File: stacks/cli/src/commands/delete.ts (hunk 4 above)
Comment (htuna07) [reply stacks/cli/src/commands/delete.ts]: I suggested `remove` word but it is okay

File: stacks/api/machinery/src/store.ts (hunk 8 above)
Comment (thesayyn) [reply stacks/api/machinery/src/store.ts]: Done

**Comment by thesayyn** [comment]
@htuna07 this PR is ready for another review.

**Comment by thesayyn** [comment]
Now that all tests are passing, we are ready to get this merged

**Review by htuna07 — DISMISSED** [review]

**Comment by TolgaYigit** [comment]
> Now that all tests are passing, we are ready to get this merged

I wonder, what happened after this message @thesayyn. 😄

**Comment by thesayyn** [comment]
> > Now that all tests are passing, we are ready to get this merged
> 
> I wonder, what happened after this message @thesayyn.

i guess new things got merged to master which has conflicts with  breaking changes introduced by this branch.

**Review by htuna07 — APPROVED** [review]
